storm-core/src/jvm/backtype/storm/windowing/WindowManager.java

```diff
@@ -0,0 +1,210 @@
+ */
+package backtype.storm.windowing;
+
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static backtype.storm.topology.base.BaseWindowedBolt.Count;
+import static backtype.storm.topology.base.BaseWindowedBolt.Duration;
+
+/**
+ * Tracks a window of events and fires {@link WindowLifecycleListener} callbacks
+ * on expiry of events or activation of the window due to {@link TriggerPolicy}.
+ *
+ * @param <T> the type of event in the window.
+ */
+public class WindowManager<T> implements TriggerHandler {
+    private static final Logger LOG = LoggerFactory.getLogger(WindowManager.class);
+
+    /**
+     * Expire old events every EXPIRE_EVENTS_THRESHOLD to
+     * keep the window size in check.
+     */
+    public static final int EXPIRE_EVENTS_THRESHOLD = 100;
+
+    private WindowLifecycleListener<T> windowLifecycleListener;
+    private ConcurrentLinkedQueue<Event<T>> window;
+    private EvictionPolicy<T> evictionPolicy;
+    private TriggerPolicy<T> triggerPolicy;
+    private List<T> expiredEvents;
+    private Set<Event<T>> prevWindowEvents;
+    private AtomicInteger eventsSinceLastExpiry;
+    private ReentrantLock lock;
+
+    public WindowManager(WindowLifecycleListener<T> lifecycleListener) {
+        windowLifecycleListener = lifecycleListener;
+        window = new ConcurrentLinkedQueue<>();
+        expiredEvents = new ArrayList<>();
+        prevWindowEvents = new HashSet<>();
+        eventsSinceLastExpiry = new AtomicInteger();
+        lock = new ReentrantLock(true);
+    }
+
+    public void setWindowLength(Count count) {
+        this.evictionPolicy = new CountEvictionPolicy<>(count.value);
+    }
+
+    public void setWindowLength(Duration duration) {
+        this.evictionPolicy = new TimeEvictionPolicy<>(duration.value);
+    }
+
+    public void setSlidingInterval(Count count) {
+        this.triggerPolicy = new CountTriggerPolicy<>(count.value, this);
+    }
+
+    public void setSlidingInterval(Duration duration) {
+        this.triggerPolicy = new TimeTriggerPolicy<>(duration.value, this);
+    }
+
+    /**
+     * Add an event into the window, with {@link System#currentTimeMillis()} as
+     * the tracking ts.
+     *
+     * @param event the event to add
+     */
+    public void add(T event) {
+        add(event, System.currentTimeMillis());
+    }
+
+    /**
+     * Add an event into the window, with the given ts as the tracking ts.
+     *
+     * @param event the event to track
+     * @param ts    the timestamp
+     */
+    public void add(T event, long ts) {
+        Event<T> windowEvent = new EventImpl<T>(event, ts);
+        window.add(windowEvent);
+        track(windowEvent);
+        compactWindow();
```

Comment:
Compaction is done to keep the window size in check, in case the trigger is defined such that unwanted events keep piling up in the window. Eg. window length = 100 and sliding interval is 1 min and the event rate is 1000 events/sec, without compaction, we will end storing upto 60,000 events whereas ideally we dont want to keep more than 100 events in the window. Similar cases can happen for other window configurations.

The events are stored in a `ConcurrentLinkedQueue` and onTrigger takes a snapshot of the queue and iterates over it even if elements are added meanwhile. There is locking to keep the compaction and expiry from stepping into each other. The idea was to minimize locking esp while adding events.